algorithms/src/std_algorithms/numeric/Kokkos_TransformReduce.hpp

```diff
@@ -0,0 +1,496 @@
+    IteratorType1 last1, IteratorType2 first2, ValueType init_reduction_value,
+    JoinerType joiner, BinaryTransformerType transformer) {
+  // checks
+  static_assert_random_access_and_accessible(ex, first1, first2);
+  static_assert_is_not_opemnptarget(ex);
+  static_assert_iterators_have_matching_difference_type(first1, first2);
+  expect_valid_range(first1, last1);
+
+  if (first1 == last1) {
+    // init is returned, unmodified
+    return init_reduction_value;
+  }
+
+  // aliases
+  using index_type = typename IteratorType1::difference_type;
+  using reducer_type =
+      ReducerWithArbitraryJoinerNoNeutralElement<ValueType, JoinerType,
+                                                 ExecutionSpace>;
+  using functor_type =
+      StdTransformReduceTwoIntervalsFunctor<index_type, IteratorType1,
+                                            IteratorType2, reducer_type,
+                                            BinaryTransformerType>;
+  using result_view_type = typename reducer_type::result_view_type;
+
+  // run
+  result_view_type result("transform_reduce_custom_functors_impl_result");
+  reducer_type reducer(result, joiner);
+
+  const auto num_elements = last1 - first1;
+  ::Kokkos::parallel_reduce(
+      label, RangePolicy<ExecutionSpace>(ex, 0, num_elements),
+      functor_type(first1, first2, reducer, transformer), reducer);
+  ex.fence("Kokkos::transform_reduce: fence after operation");
+
+  // return
+  const auto r_h =
+      ::Kokkos::create_mirror_view_and_copy(::Kokkos::HostSpace(), result);
+  return joiner(r_h().val, init_reduction_value);
+}
+
+template <class ExecutionSpace, class IteratorType1, class IteratorType2,
+          class ValueType>
+ValueType transform_reduce_default_functors_impl(
+    const std::string& label, const ExecutionSpace& ex, IteratorType1 first1,
+    IteratorType1 last1, IteratorType2 first2, ValueType init_reduction_value) {
+  // checks
+  static_assert_random_access_and_accessible(ex, first1, first2);
+  static_assert_is_not_opemnptarget(ex);
+  static_assert_iterators_have_matching_difference_type(first1, first2);
+  expect_valid_range(first1, last1);
+
+  // aliases
+  using transformer_type =
+      Impl::StdTranformReduceDefaultBinaryTransformFunctor<ValueType>;
+  using joiner_type = Impl::StdTranformReduceDefaultJoinFunctor<ValueType>;
+
+  return transform_reduce_custom_functors_impl(
+      label, ex, first1, last1, first2, std::move(init_reduction_value),
+      joiner_type(), transformer_type());
+}
+
+}  // end namespace Impl
+
+///////////////////////////////
+//
+// transform_reduce public API
+//
+///////////////////////////////
+
+// ----------------------------
+// overload set1:
+// no custom functors passed, so equivalent to
+// transform_reduce(first1, last1, first2, init, plus<>(), multiplies<>());
+// ----------------------------
+template <class ExecutionSpace, class IteratorType1, class IteratorType2,
```

Comment:
This applies in other places where the API is separated from the `*_impl` (which I think is really good design), but maybe we should add SFINAE/requires (when we eventually move to C++20 concepts) checking the traits of these iterators so we get an earlier error on the overload resolution rather than as a static assert when the `*_impl` function checks the type

Reply:
agree about the require thing. All the impl we have was assuming we can use C++14 but nothing more recent. A lot of things can be improved with more recent standard.